packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/Relations/RelationModal.tsx

```diff
@@ -57,91 +59,193 @@ const CustomModalContent = styled(Modal.Content)`
+  rootDocumentMeta: DocumentMeta;
+  currentDocumentMeta: DocumentMeta;
+  currentDocument: ReturnType<UseDocument>;
+  onPreview?: () => void;
+}
 
-  const [triggerRefetchDocument] = useLazyGetDocumentQuery();
+const [RelationModalProvider, useRelationModal] =
+  createContext<RelationModalContextValue>('RelationModal');
 
-  const currentDocument = useDocumentContext('RelationModalForm', (state) => state.document);
-  const rootDocumentMeta = useDocumentContext(
-    'RelationModalForm',
-    (state) => state.rootDocumentMeta
-  );
-  const currentDocumentMeta = useDocumentContext('RelationModalForm', (state) => state.meta);
-  const changeDocument = useDocumentContext('RelationModalForm', (state) => state.changeDocument);
-  const documentHistory = useDocumentContext('RelationModalForm', (state) => state.documentHistory);
-  const setDocumentHistory = useDocumentContext(
-    'RelationModalForm',
-    (state) => state.setDocumentHistory
-  );
+const RelationContextWrapper = ({ children }: { children: React.ReactNode }) => {
+  const rootDocument = useDoc();
+
+  const [state, dispatch] = React.useReducer(reducer, {
+    documentHistory: [],
+    confirmDialogIntent: null,
+    isModalOpen: false,
+  });
+
+  if (!rootDocument.document) {
+    throw new Error('Root document not found');
+  }
+
+  const rootDocumentMeta: DocumentMeta = {
+    documentId: rootDocument.document.documentId,
+    model: rootDocument.model,
+    collectionType: rootDocument.collectionType,
+  };
+
+  const currentDocumentMeta = state.documentHistory.at(-1) ?? rootDocumentMeta;
 
-  const [isConfirmationOpen, setIsConfirmationOpen] = React.useState(false);
-  const [actionPosition, setActionPosition] = React.useState<'cancel' | 'back' | 'navigate'>(
-    'cancel'
+  const params = React.useMemo(
+    () => buildValidParams(currentDocumentMeta.params ?? {}),
+    [currentDocumentMeta.params]
   );
+  const currentDocument = useDocument({ ...currentDocumentMeta, params });
 
-  const [isModalOpen, setIsModalOpen] = React.useState(false);
-  // NOTE: Not sure about this relation modal context, maybe we should move this to DocumentContext?
-  // Get parent modal context if it exists
-  const parentContext = useRelationModal('RelationModalForm', (state) => state);
-  // Get depth of nested modals
-  const depth = parentContext ? parentContext.depth + 1 : 0;
-  // Check if this is a nested modal
-  const isNested = depth > 0;
+  const memoizedChildren = React.useMemo(() => children, [children]);
 
-  const addDocumentToHistory = (document: DocumentMeta) =>
-    setDocumentHistory((prev) => [...prev, document]);
+  // Return children directly if the context is already in the tree
+  try {
+    useRelationModal('RelationModalWrapper', () => null);
+    // If we reach that point without an error, the context is already in the tree
+    return memoizedChildren;
+  } catch (e) {
+    // Silently catch the error. We will wrap children in the context provider below
+  }
```

Comment:
This isn't quite working yet. I'm getting an endless rerender issue that crashes my browser

Reply:
I had another idea on my run earlier. If this current approach turns out to be tricky, we could also decide to just recursively wrap the context inside itself anyway. I think it should work if we always forward the context value from the parent

```js
const context = useRelationModal(state => state);

return <RelationModalProvider {...state}>{children}</RelationModalProvider>
```

useContext just matches the closest parent, it doesn't complain about being nested inside itself, so I believe eventually it should resolve to the same parent context so everything is still managed in the same reducer